Align URLPattern path canonicalisation to spec

youennf opened the pull request:
#### 8328992f273c7ac81d6c3f7180fb8d68154131f0
<pre>
Align URLPattern path canonicalisation to spec
<a href="https://bugs.webkit.org/show_bug.cgi?id=285982">https://bugs.webkit.org/show_bug.cgi?id=285982</a>
<a href="https://rdar.apple.com/problem/142952620">rdar://problem/142952620</a>

Reviewed by Anne van Kesteren.

An empty scheme should trigger pathname canonicalization and not opaque pathname canonicalization,
as per <a href="https://urlpattern.spec.whatwg.org/#process-pathname-for-init">https://urlpattern.spec.whatwg.org/#process-pathname-for-init</a> step 2.

In addition, to disable percent encoding of pathname for opaque path names, we use the URL parser via URL constructor and not via URL::setPath.
And we use a non special scheme as part of it.

* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.https.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.https.any.worker-expected.txt:
* Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
(WebCore::canonicalizeOpaquePathname):
(WebCore::processPathname):

Canonical link: <a href="https://commits.webkit.org/288934@main">https://commits.webkit.org/288934@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f6b2d167071bc9817c1adf24544cf7bd0c729ecc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/84836 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/4561 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/39224 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/89975 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/35888 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/86921 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/4650 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/12538 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/66025 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/23842 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/87881 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/3544 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/77096 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/46295 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/3423 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/31320 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/34962 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/74248 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/32128 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/91351 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/12175 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/8892 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/74503 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/12402 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/72906 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/73626 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18214 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/18002 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/16453 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/3626 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/12127 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/17573 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/11961 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/15461 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/13707 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
```diff
@@ -192,23 +192,13 @@ ExceptionOr<String> canonicalizeOpaquePathname(StringView value)
-    // Prepend dash to avoid inadvertantly collapsing a leading dot due to the fake leading slash.
-    String maybeAddSlashPrefix = hasLeadingSlash ? value.toString() : makeString("/-"_s, value);
-
     // FIXME: Set state override to State::OpaquePath after URLParser supports state override.
```
Maybe we can remove this FIXME as well? I don't think we really support state overrides currently and this implementation seems reasonably solid.

youennf replied:
Done

youennf committed:
Align URLPattern path canonicalisation to spec
https://bugs.webkit.org/show_bug.cgi?id=285982
rdar://problem/142952620

Reviewed by Anne van Kesteren.

An empty scheme should trigger pathname canonicalization and not opaque pathname canonicalization,
as per https://urlpattern.spec.whatwg.org/#process-pathname-for-init step 2.

In addition, to disable percent encoding of pathname for opaque path names, we use the URL parser via URL constructor and not via URL::setPath.
And we use a non special scheme as part of it.

* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.https.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.https.any.worker-expected.txt:
* Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
(WebCore::canonicalizeOpaquePathname):
(WebCore::processPathname):

Canonical link: https://commits.webkit.org/288934@main

jelee53 commented on Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
```diff
@@ -195,23 +195,12 @@ ExceptionOr<String> canonicalizeOpaquePathname(StringView value)
-    // FIXME: Set state override to State::OpaquePath after URLParser supports state override.
-    URL dummyURL(dummyURLCharacters);
-    dummyURL.setPath(maybeAddSlashPrefix);
+    URL dummyURL(makeString("a:"_s, value));
```
Why do we have to use a non-special scheme when constructing the URL? Is that step also necessary to disable to percent encoding?

Also, I know that this has already been committed by why not add an existing non-special scheme like `rdar:` or `file` or something along those lines? Otherwise , people might be confused as to what "a:" does?
